traits/observers/_observer_change_notifier.py

```diff
@@ -0,0 +1,183 @@
+        self.observer_handler = observer_handler
+        self.event_factory = event_factory
+        self.prevent_event = prevent_event
+        self.graph = graph
+        self.target = weakref.ref(target)
+        if isinstance(handler, types.MethodType):
+            self.handler = weakref.WeakMethod(handler)
+        else:
+            self.handler = partial(_return, value=handler)
+        self.dispatcher = dispatcher
+
+    def add_to(self, observable):
+        """ Add this notifier to the observable.
+
+        Parameters
+        ----------
+        observable : IObservable
+        """
+        notifiers = observable._notifiers(True)
+        notifiers.append(self)
+
+    def remove_from(self, observable):
+        """ Remove a notifier equivalent to this one from the observable.
+
+        Parameters
+        ----------
+        observable : IObservable
+
+        Raises
+        ------
+        NotifierNotFound
+            If the notifier cannot be found.
+        """
+        notifiers = observable._notifiers(True)
+        for notifier in notifiers[:]:
+            if self.equals(notifier):
+                notifiers.remove(notifier)
+                break
+        else:
+            raise NotifierNotFound("Notifier not found.")
+
+    def __call__(self, *args, **kwargs):
+        """ Called by the observable this notifier is attached to.
+
+        This exercises the observer_handler for maintaining observers and
+        notifiers on changed objects.
+
+        Note that any unexpected exceptions will be raised, as the
+        ``observer_handler`` is not provided by users of traits, but is
+        a callable maintained in traits.
+        """
+        target = self.target()
+        if target is None:
+            return
+
+        handler = self.handler()
+        if handler is None:
+            return
+
+        event = self.event_factory(*args, **kwargs)
+        if self.prevent_event(event):
+            return
+
+        self.observer_handler(
+            event=event,
+            graph=self.graph,
+            target=target,
+            handler=handler,
+            dispatcher=self.dispatcher,
+        )
+
+    def equals(self, other):
+        """ Return true if the other value is a notifier equivalent to this one.
+
+        Parameters
+        ----------
+        other : any
+
+        Returns
+        -------
+        boolean
+        """
+        return (
+            type(self) is type(other)
+            # observer_handler contains the logic for maintaining notifiers
+            # in the downstream graph.
+            and self.observer_handler is other.observer_handler
+            # graph is an input for observer_handler.
+            # Unequal graphs should not interfere each other.
+            and self.graph == other.graph
+            # user handler is an input for observer_handler.
+            # different user handlers should not interfere each other.
+            and self.handler() == other.handler()
```

Comment:
This has the possibility of spurious equality in the case where:

- everything is the same _except_ the handler
- you have two different method handlers
- which have both been garbage collected

In this case both will return `None` and compare equal.  Of course in this situation they could both be `None` because it was the _same_ handler which had been gc'd.

You should try to do something sensible if `handler()` returns None, which might be to raise, or it may be something else.  It may even be that this is OK as-is.

Reply:
If both `self.handler()` and `other.handler()` are initially referring to different instance methods and the instances have been garbage collected, and everything else is the same, then the two notifiers are genuinely indistinguishable from one and other. These notifiers are silenced too.
I think overall not dealing with `handler() is None` is okay...I can't think of a test that would fail where the test represents some actual use case.